src/main/java/hudson/plugins/promoted_builds/integrations/jobdsl/PromotionContext.java

```diff
@@ -1,108 +1,145 @@
 package hudson.plugins.promoted_builds.integrations.jobdsl;
 
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
 import groovy.lang.Closure;
+import groovy.lang.MetaClass;
 import groovy.util.Node;
-import hudson.plugins.promoted_builds.PromotionCondition;
-import hudson.plugins.promoted_builds.conditions.DownstreamPassCondition;
-import hudson.plugins.promoted_builds.conditions.ParameterizedSelfPromotionCondition;
-import hudson.plugins.promoted_builds.conditions.SelfPromotionCondition;
-import hudson.plugins.promoted_builds.conditions.UpstreamPromotionCondition;
+import groovy.util.NodeBuilder;
+import groovy.util.NodeList;
+import javaposse.jobdsl.dsl.DslContext;
+import javaposse.jobdsl.dsl.Item;
+import javaposse.jobdsl.dsl.JobManagement;
+import javaposse.jobdsl.dsl.helpers.step.StepContext;
+import javaposse.jobdsl.plugin.DslEnvironment;
+import org.codehaus.groovy.runtime.InvokerHelper;
+import org.codehaus.groovy.runtime.MethodClosure;
 
 import java.util.ArrayList;
 import java.util.List;
 
-import javaposse.jobdsl.dsl.Context;
-import javaposse.jobdsl.dsl.helpers.step.StepContext;
-
-import javaposse.jobdsl.plugin.DslEnvironment;
-
 import static javaposse.jobdsl.plugin.ContextExtensionPoint.executeInContext;
 
-class PromotionContext implements Context {
-    private final DslEnvironment dslEnvironment;
+/**
+ * Implementation notes: in order to get advanced integration with Job DSL which is Groovy-powered, we make use of
+ * some helper classes such as the {@link MethodClosure} adapter used for methods accepting Closure arguments. This may
+ * result in some unusual OOP constructs.
+ */
+class PromotionContext extends Item {
```

Comment:
The new implementation relies on 2 aspects of the Job DSL plugin that are currently undocumented in the aforementioned guideline:
- ability to tap into the automatic DSL engine via `@ContextType` / providing our ___own___ `ExtensibleContext` for [JENKINS-39342](https://issues.jenkins-ci.org/browse/JENKINS-39342)
- ability to tap into the XML generation engine via `Item` extension for [JENKINS-34982](https://issues.jenkins-ci.org/browse/JENKINS-34982) for API consistency with the core DSL (a promotion has its own XML configuration file and `Item` contains access to the XML generation tooling)

Can the DSL extension guideline be reviewed to address those needs? If we cannot, this DSL extension will have to implements its own alternative to `DescribableHelper`/`DescribableContext`/`DescribableListContext` (automatic DSL extension), `ContextHelper`/`NodeEnhancement`/`MissingPropertyToStringDelegate` which is far behind a simple integration with Job DSL, that would be alternative DSL implementation (within the scope of the extension) which may not provide the same user experience as the core DSL.